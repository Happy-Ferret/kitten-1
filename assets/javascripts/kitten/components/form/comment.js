import React, { Component, Fragment } from 'react'
import Radium, { StyleRoot } from 'radium'
import PropTypes from 'prop-types'
import { Marger as MargerBase } from 'kitten/components/layout/marger'
import { CommentAvatar } from 'kitten/components/form/comment-avatar'
import { Text } from 'kitten/components/typography/text'
import { ScreenConfig } from 'kitten/constants/screen-config'
import COLORS from 'kitten/constants/colors-config'

const Marger = Radium(MargerBase)

export class Comment extends Component {
  static propTypes = {
    text: PropTypes.string.isRequired,
    ownerName: PropTypes.string.isRequired,
    avatarImgProps: PropTypes.object.isRequired,
    commentDate: PropTypes.string.isRequired,
  }

  render() {
    const { avatarImgProps, commentDate } = this.props

    return (
      <StyleRoot>
        <div style={styles.grid}>
          <CommentAvatar
            avatarImgProps={avatarImgProps}
            commentDate={commentDate}
          />
          {this.renderComment()}
        </div>
      </StyleRoot>
    )
  }

  renderComment() {
    const { text, ownerName } = this.props

    return (
      <div style={styles.comment}>
        <Marger bottom="1">
          <Text color="font1" size="tiny" weight="regular">
            {ownerName}
          </Text>
        </Marger>
        <Marger top="1">
          <Text color="font1" size="tiny" weight="light">
            {text}
          </Text>
        </Marger>
        <span style={styles.comment.arrow} />
      </div>
    )
  }
}

const styles = {
  grid: {
    display: 'flex',
  },
  comment: {
    position: 'relative',
    borderWidth: '2px',
    backgroundColor: COLORS.background3,
    borderColor: COLORS.background3,
    color: COLORS.font1,
    padding: '30px',
    fontSize: '16px',
    marginLeft: 20,
    [`@media (min-width: ${ScreenConfig['S'].min}px)`]: {
      marginLeft: 35,
    },

    arrow: {
      position: 'absolute',
      top: '20px',
      display: 'block',
      width: 0,
      height: 0,
      borderWidth: '10px',
      borderStyle: 'solid',
      borderColor: 'transparent',
      borderRightColor: COLORS.background3,
      left: '-20px',
      [`@media (min-width: ${ScreenConfig['S'].min}px)`]: {
        top: '35px',
      },
    },
  },
}
