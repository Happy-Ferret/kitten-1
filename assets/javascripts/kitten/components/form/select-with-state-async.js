// Select that uses https://github.com/JedWatson/react-select
// with default states and handles the current value as a local state.
//
// Also handles multi-level options. You can give `children: []` to each option
// and fill it with options to represent a hierarchy.
//
// This version of the component uses asynchronous loading of options
// instead of a predefined list of values.
//
//
// E.g:
//    <SelectWithStateAsync loadOptions={} />
//
import React from 'react'
import classNames from 'classnames'
import Select from 'react-select'

class SelectWithStateAsync extends React.Component {
  constructor(props) {
    super(props)

    this.state = {
      value: props.value,
    }

    this.handleChange = this.handleChange.bind(this)
  }

  handleChange(val) {
    this.setState({ value: val })
    this.props.onChange(val)
    this.props.onInputChange({ value: val, name: this.props.name })
  }

  renderLabel() {
    if (!this.props.labelText) return ''

    return (
      <label className="k-Select__label" id={ this.props.id }>
        { this.props.labelText }
      </label>
    )
  }

  render() {
    const { value, onChange, className, tiny, error, valid, disabled, ...other } = this.props
    const selectClassName = classNames(
      'k-Select',
      className,
      {
        'k-Select--tiny': tiny,
        'is-error': error,
        'is-valid': valid,
        'is-disabled': disabled,
      }
    )

    return (
      <div className={ selectClassName }>
        { this.renderLabel() }
        <SelectWithMultiLevelAsync value={ this.state.value }
                                   onChange={ this.handleChange }
                                   disabled={ disabled }
                                   { ...other } />
      </div>
    )
  }
}

class SelectWithMultiLevelAsync extends React.Component {
  // Turns a hierarchy of options with `children` into a flat array
  // of options with a `level` of 1, 2 or null.
  // TODO: Where should we post-process the list of returned options?
  flattenedOptions() {
    const options = []

    this.props.options.map(option => {
      if (option.children) {
        option.level = 1
        options.push(option)
        option.children.map(opt => {
          opt.level = 2
          options.push(opt)
        })
      } else {
        options.push(option)
      }
    })

    return options
  }

  // React-Select allows changing the way options are rendered.
  optionRenderer({ level, label }) {
    if (!level) return label

    return (
      <span className={ 'k-Select__option--level' + level }>
        { label }
      </span>
    )
  }

  render() {
    let inputProps = this.props.inputProps

    if (this.props.labelText && !inputProps['aria-labelledby'])
      inputProps['aria-labelledby'] = this.props.id

    return <Select.Async optionRenderer={ this.optionRenderer }
                         { ...this.props }
                         inputProps={ inputProps }
                         options={ this.props.options }
                         loadOptions={ this.props.loadOptions } />
  }
}

SelectWithStateAsync.defaultProps = {
  onChange: function() {},
  onInputChange: function() {},
  clearable: false,
  searchable: false,
  deleteRemoves: false,
  multi: false,
  labelText: null,
  error: false,
  valid: false,
  disabled: false,
  tiny: false,
  name: null,
  inputProps: {},
}

export default SelectWithStateAsync
