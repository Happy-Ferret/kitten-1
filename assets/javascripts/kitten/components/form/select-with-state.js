// Select that uses https://github.com/JedWatson/react-select
// with default states and handles the current value as a local state.
//
// Also handles multi-level options. You can give `children: []` to each option
// and fill it with options to represent a hierarchy.
//
// E.g:
//    <SelectWithState options={
//      { label: 'Cat',
//        children: [{ label: 'A' }, { label: 'B' }]
//      } />

import React from 'react'
import classNames from 'classnames'
import Select from 'react-select'

class SelectWithState extends React.Component {
  constructor(props) {
    super(props)

    this.state = {
      value: props.value,
    }

    this.handleChange = this.handleChange.bind(this)
  }

  handleChange(val) {
    this.setState({ value: val })
    this.props.onChange(val)
    this.props.onInputChange({ value: val, name: this.props.name })
  }

  renderLabel() {
    if (!this.props.labelText) return ''

    return (
      <label className="k-Select__label" id={ this.props.id }>
        { this.props.labelText }
      </label>
    )
  }

  render() {
    const {
      value,
      onChange,
      className,
      tiny,
      error,
      valid,
      disabled,
      ...other,
    } = this.props

    const selectClassName = classNames(
      'k-Select',
      className,
      {
        'k-Select--tiny': tiny,
        'is-error': error,
        'is-valid': valid,
        'is-disabled': disabled,
      }
    )

    return (
      <div className={ selectClassName }>
        { this.renderLabel() }
        <SelectWithMultiLevel value={ this.state.value }
                              onChange={ this.handleChange }
                              disabled={ disabled }
                              { ...other } />
      </div>
    )
  }
}

class SelectWithMultiLevel extends React.Component {
  constructor(props) {
    super(props)

    this.loadAndFlattenOptions = this.loadAndFlattenOptions.bind(this)
  }

  // Turns a hierarchy of options with `children` into a flat array
  // of options with a `level` of 1, 2 or null.
  flattenedOptions(rawOptions = this.props.options) {
    const options = []

    rawOptions.map(option => {
      if (option.children) {
        option.level = 1
        options.push(option)
        option.children.map(opt => {
          opt.level = 2
          options.push(opt)
        })
      } else {
        options.push(option)
      }
    })

    return options
  }

  loadAndFlattenOptions(input, callback) {
    this.props.loadOptions(
      input,
      (error, options) => {
        if (!error) {
          callback(error, { options: this.flattenOptions(options) })
        } else {
          callback(error, { options })
        }
      }
    )
  }

  // React-Select allows changing the way options are rendered.
  optionRenderer({ level, label }) {
    if (!level) return label

    return (
      <span className={ `k-Select__option--level${ level }` }>
        { label }
      </span>
    )
  }

  render() {
    let inputProps = this.props.inputProps

    if (this.props.labelText && !inputProps['aria-labelledby']) {
      inputProps['aria-labelledby'] = this.props.id
    }

    const Component = this.props.async ? Select.Async : Select

    return <Component
      { ...this.props }
      optionRenderer={ this.optionRenderer }
      inputProps={ inputProps }
      options={ this.flattenedOptions() }
      loadOptions={ this.loadAndFlattenOptions }
    />
  }
}

SelectWithState.defaultProps = {
  onChange: () => {},
  onInputChange: () => {},
  loadOptions: () => {},
  clearable: false,
  searchable: false,
  deleteRemoves: false,
  multi: false,
  labelText: null,
  error: false,
  valid: false,
  disabled: false,
  tiny: false,
  name: null,
  inputProps: {},
  async: false,
}

export default SelectWithState
