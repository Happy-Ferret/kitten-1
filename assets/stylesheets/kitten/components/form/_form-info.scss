/// Form info
///
/// A form info element that can take input/checkbox/radio button etc. element
/// with an error message.
///
/// @group form
///
/// @parameter {Map} $input - a input specification
///
/// @example scss - Usage
///
/// @include k-FormInfo;
///
/// <div class="k-FormInfo">
///   <label class="k-FormInfo__label">Nom du label</label>
///   <input class="k-TextInput"
///          type="text"
///          name=""
///          placeholder="Lorem ipsum dolor sit amet" />
/// </div>

@mixin k-FormInfo {
  // Label
  $font: 'regular';
  $font-size-label: 0; // 16px
  $color-label: map-get($k-colors, 'font-1');

  // Error
  $font-size-error: -2; // 12px
  $color-error: map-get($k-colors, 'error');

  .k-FormInfo__label {
    @include k-typographyFont($font);
    @include k-typographyFontSize($font-size-label);

    display: block;
    padding-bottom: k-px-to-rem(10px);

    color: $color-label;
  }

  .k-FormInfo__error {
    @include k-typographyFont($font);
    @include k-typographyFontSize($font-size-error);

    margin-top: k-px-to-rem(5px);

    color: $color-error;
  }
}
