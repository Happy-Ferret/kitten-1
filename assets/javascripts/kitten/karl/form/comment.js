import React from 'react'
import Radium, { StyleRoot } from 'radium'
import { Comment } from 'kitten/components/form/comment'
import { CommentForm } from 'kitten/components/form/comment-form'
import { Container } from 'kitten/components/grid/container'
import { Marger } from 'kitten/components/layout/marger'
import { KarlExampleTitle } from 'kitten/karl/examples/title'

export const KarlComment = () => (
  <Fragment>
    <KarlExampleTitle>Comment</KarlExampleTitle>
    <Container>
      <Marger top="2" bottom="2">
        <Comment
          text="Lorem ipsum dolor sit amet, consectetur adipiscing elit, sed do eiusmod tempor incididunt ut labore et dolore magna aliqua. Ut enim ad minim veniam, quis nostrud exercitation ullamco laboris"
          ownerName="Lorem ipsum"
        />
      </Marger>
    </Container>

    <KarlExampleTitle>Comment small text</KarlExampleTitle>
    <Container>
      <Marger top="2" bottom="2">
        <Comment text="Foo Bar" ownerName="Lorem ipsum" />
      </Marger>
    </Container>
  </Fragment>
)

export const KarlCommentForm = () => (
  <Fragment>
    <KarlExampleTitle>Comment form</KarlExampleTitle>
    <Container>
      <Marger top="2" bottom="2">
        <CommentForm
          ownerImgProps={(src = 'https://placehold.it/80x80/caf4fe/caf4fe')}
          ownerTime="2 min"
          placeholder="Leave a comment on the project"
          commentButton="Comment"
        />
      </Marger>
    </Container>

    <KarlExampleTitle>Comment form error</KarlExampleTitle>
    <Container>
      <Marger top="2" bottom="2">
        <CommentForm text="Foo Bar" ownerName="Lorem ipsum" />
      </Marger>
    </Container>
  </Fragment>
)
