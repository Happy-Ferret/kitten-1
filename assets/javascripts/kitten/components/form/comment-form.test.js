import React from 'react'
import renderer from 'react-test-renderer'
import { CommentForm } from 'kitten/components/form/comment-form'

const createMockMediaMatcher = matches => () => ({
  matches,
  addListener: () => {},
  removeListener: () => {},
})

describe('<CommentForm />', () => {
  let originalMatchMedia
  let component

  beforeEach(() => {
    originalMatchMedia = window.matchMedia
  })

  afterEach(() => {
    window.matchMedia = originalMatchMedia
  })

  it('should match its empty snapshot', () => {
    window.matchMedia = createMockMediaMatcher(false)

    const tree = renderer
      .create(
        <CommentForm
          avatarImgProps={{
            src: 'https://placehold.it/80x80/caf4fe/caf4fe',
            alt: '',
          }}
          placeholder="Leave a comment on the project"
          commentButton="Comment"
          errorMessage="Cum sociis natoque penatibus et magnis"
          error
        />,
      )
      .toJSON()

    expect(tree).toMatchSnapshot()
  })
})
