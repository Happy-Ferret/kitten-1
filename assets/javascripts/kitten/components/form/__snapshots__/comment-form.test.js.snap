// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`<CommentForm /> should match its empty snapshot 1`] = `
<div
  data-radium={true}
>
  <div
    data-radium={true}
    style={
      Object {
        "display": "flex",
      }
    }
  >
    <div
      data-radium={true}
      style={
        Object {
          "alignItems": "center",
          "display": "flex",
          "flexDirection": "column",
        }
      }
    >
      <div
        data-radium={true}
        style={
          Object {
            "marginBottom": "0.625rem",
            "marginTop": null,
          }
        }
      >
        <span
          className="k-ButtonImage k-ButtonImage--huge k-ButtonImage--withoutPointerEvents"
          tabIndex={null}
        >
          <img
            alt=""
            className="k-ButtonImage__img"
            src="https://placehold.it/80x80/caf4fe/caf4fe"
          />
        </span>
      </div>
      <span
        className="k-u-color-font1 k-u-size-micro k-u-weight-light"
      >
        
      </span>
    </div>
    <style
      dangerouslySetInnerHTML={
        Object {
          "__html": ".k-CommentForm__input::-webkit-input-placeholder{color: #b8b8b8;}",
        }
      }
    />
    <style
      dangerouslySetInnerHTML={
        Object {
          "__html": ".k-CommentForm__input:-moz-placeholder{color: #b8b8b8;}",
        }
      }
    />
    <style
      dangerouslySetInnerHTML={
        Object {
          "__html": ".k-CommentForm__input::-moz-placeholder{color: #b8b8b8;}",
        }
      }
    />
    <style
      dangerouslySetInnerHTML={
        Object {
          "__html": ".k-CommentForm__input:-ms-input-placeholder{color: #b8b8b8;}",
        }
      }
    />
    <div
      className="rmq-993edb07"
      data-radium={true}
      style={
        Object {
          "flex": 1,
          "marginLeft": 20,
        }
      }
    >
      <div
        data-radium={true}
        style={
          Object {
            "display": "flex",
            "marginBottom": "0.3125rem",
            "marginTop": null,
            "position": "relative",
          }
        }
      >
        <textarea
          className="k-CommentForm__input"
          data-radium={true}
          disabled={false}
          onBlur={[Function]}
          onChange={[Function]}
          onFocus={[Function]}
          placeholder="Leave a comment on the project"
          rows="3"
          style={
            Object {
              "borderColor": "#ffb2c7",
              "borderStyle": "solid",
              "borderWidth": "2px",
              "color": "#ffb2c7",
              "fontFamily": "Maax, Helvetica, Arial, sans-serif",
              "fontSize": "16px",
              "padding": "30px",
              "width": "100%",
            }
          }
        >
          
        </textarea>
        <span
          className="rmq-e4f3f344"
          data-radium={true}
          style={
            Object {
              "borderColor": "transparent",
              "borderRightColor": "#ffb2c7",
              "borderStyle": "solid",
              "borderWidth": "10px",
              "display": "block",
              "height": 0,
              "left": "-20px",
              "position": "absolute",
              "top": "20px",
              "width": 0,
            }
          }
        >
          <span
            data-radium={true}
            style={
              Object {
                "borderColor": "transparent",
                "borderRightColor": "white",
                "borderStyle": "solid",
                "borderWidth": "10px",
                "height": 0,
                "left": "-7px",
                "marginTop": "-10px",
                "position": "absolute",
                "width": 0,
              }
            }
          />
        </span>
      </div>
      <div
        data-radium={true}
        style={
          Object {
            "marginBottom": null,
            "marginTop": "0.3125rem",
          }
        }
      >
        <span
          className="k-u-color-error k-u-size-micro k-u-weight-regular"
        >
          Cum sociis natoque penatibus et magnis
        </span>
      </div>
    </div>
  </div>
  <style
    dangerouslySetInnerHTML={
      Object {
        "__html": "@media (min-width: 640px){ .rmq-993edb07{margin-left: 35px !important;}}
@media (min-width: 640px){ .rmq-e4f3f344{top: 35px !important;}}",
      }
    }
  />
</div>
`;
